<?xml version="1.0" encoding="ascii"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
          "DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
  <title>ifr.interfaces.gather.gather</title>
  <link rel="stylesheet" href="epydoc.css" type="text/css" />
  <script type="text/javascript" src="epydoc.js"></script>
</head>

<body bgcolor="white" text="black" link="blue" vlink="#204080"
      alink="#204080">
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="ifr-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="module-tree.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            >Iterative Feature Removal</th>
          </tr></table></th>
  </tr>
</table>
<table width="100%" cellpadding="0" cellspacing="0">
  <tr valign="top">
    <td width="100%">
      <span class="breadcrumbs">
        <a href="ifr-module.html">Package&nbsp;ifr</a> ::
        <a href="ifr.interfaces-module.html">Package&nbsp;interfaces</a> ::
        <a href="ifr.interfaces.gather-module.html">Module&nbsp;gather</a> ::
        Class&nbsp;gather
      </span>
    </td>
    <td>
      <table cellpadding="0" cellspacing="0">
        <!-- hide/show private -->
        <tr><td align="right"><span class="options">[<a href="javascript:void(0);" class="privatelink"
    onclick="toggle_private();">hide&nbsp;private</a>]</span></td></tr>
        <tr><td align="right"><span class="options"
            >[<a href="frames.html" target="_top">frames</a
            >]&nbsp;|&nbsp;<a href="ifr.interfaces.gather.gather-class.html"
            target="_top">no&nbsp;frames</a>]</span></td></tr>
      </table>
    </td>
  </tr>
</table>
<!-- ==================== CLASS DESCRIPTION ==================== -->
<h1 class="epydoc">Class gather</h1><p class="nomargin-top"><span class="codelink"><a href="ifr.interfaces.gather-pysrc.html#gather">source&nbsp;code</a></span></p>
<pre class="base-tree">
object --+
         |
        <strong class="uidshort">gather</strong>
</pre>

<hr />
<!-- ==================== INSTANCE METHODS ==================== -->
<a name="section-InstanceMethods"></a>
<table class="summary" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Instance Methods</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-InstanceMethods"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ifr.interfaces.gather.gather-class.html#__init__" class="summary-sig-name">__init__</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">url</span>=<span class="summary-sig-default"><code class="variable-quote">'</code><code class="variable-string">http://gather.genome.duke.edu</code><code class="variable-quote">'</code></span>)</span><br />
      x.__init__(...) initializes x; see help(type(x)) for signature</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ifr.interfaces.gather-pysrc.html#gather.__init__">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ifr.interfaces.gather.gather-class.html#query" class="summary-sig-name">query</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">genelist</span>,
        <span class="summary-sig-arg">homologs</span>=<span class="summary-sig-default">True</span>)</span><br />
      Processes the query for the given list of genes.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ifr.interfaces.gather-pysrc.html#gather.query">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ifr.interfaces.gather.gather-class.html#getDataForAnnotation" class="summary-sig-name">getDataForAnnotation</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">key</span>=<span class="summary-sig-default"><code class="variable-quote">'</code><code class="variable-string">GO:0009615</code><code class="variable-quote">'</code></span>,
        <span class="summary-sig-arg">return_mult</span>=<span class="summary-sig-default">True</span>,
        <span class="summary-sig-arg">bf_thresh</span>=<span class="summary-sig-default">0.0</span>)</span><br />
      Returns the parsed data for all rows which contain a given key, which
      can be any string that one expects to find in the response data from 
      a gather query.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ifr.interfaces.gather-pysrc.html#gather.getDataForAnnotation">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="getRowIdsForAnnotation"></a><span class="summary-sig-name">getRowIdsForAnnotation</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">key</span>=<span class="summary-sig-default"><code class="variable-quote">'</code><code class="variable-string">virus</code><code class="variable-quote">'</code></span>)</span><br />
      Returns all the row ids for the current query results where the data 
      contains the string provided by key.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ifr.interfaces.gather-pysrc.html#gather.getRowIdsForAnnotation">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ifr.interfaces.gather.gather-class.html#getBayesFactor" class="summary-sig-name">getBayesFactor</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">row_id</span>)</span><br />
      Given the gather row identifier (not the index of self.data, but 
      rather the first field of a response row in the data), returns the 
      bayes factor and p-value.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ifr.interfaces.gather-pysrc.html#gather.getBayesFactor">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ifr.interfaces.gather.gather-class.html#parseAnnotation" class="summary-sig-name">parseAnnotation</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">annotationList</span>)</span><br />
      Parses an annotation to generate a dictionary with key-value 
      bindings.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ifr.interfaces.gather-pysrc.html#gather.parseAnnotation">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="countRows"></a><span class="summary-sig-name">countRows</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">bf_thresh</span>=<span class="summary-sig-default">0</span>)</span><br />
      counts the number of data rows (annotations) that have a bayes factor
      score &gt;= bf_thresh</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ifr.interfaces.gather-pysrc.html#gather.countRows">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
  <tr>
    <td colspan="2" class="summary">
    <p class="indent-wrapped-lines"><b>Inherited from <code>object</code></b>:
      <code>__delattr__</code>,
      <code>__format__</code>,
      <code>__getattribute__</code>,
      <code>__hash__</code>,
      <code>__new__</code>,
      <code>__reduce__</code>,
      <code>__reduce_ex__</code>,
      <code>__repr__</code>,
      <code>__setattr__</code>,
      <code>__sizeof__</code>,
      <code>__str__</code>,
      <code>__subclasshook__</code>
      </p>
    </td>
  </tr>
</table>
<!-- ==================== PROPERTIES ==================== -->
<a name="section-Properties"></a>
<table class="summary" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Properties</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-Properties"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
  <tr>
    <td colspan="2" class="summary">
    <p class="indent-wrapped-lines"><b>Inherited from <code>object</code></b>:
      <code>__class__</code>
      </p>
    </td>
  </tr>
</table>
<!-- ==================== METHOD DETAILS ==================== -->
<a name="section-MethodDetails"></a>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Method Details</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-MethodDetails"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
</table>
<a name="__init__"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">__init__</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">url</span>=<span class="sig-default"><code class="variable-quote">'</code><code class="variable-string">http://gather.genome.duke.edu</code><code class="variable-quote">'</code></span>)</span>
    <br /><em class="fname">(Constructor)</em>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ifr.interfaces.gather-pysrc.html#gather.__init__">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>x.__init__(...) initializes x; see help(type(x)) for signature</p>
  <dl class="fields">
    <dt>Overrides:
        object.__init__
        <dd><em class="note">(inherited documentation)</em></dd>
    </dt>
  </dl>
</td></tr></table>
</div>
<a name="query"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">query</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">genelist</span>,
        <span class="sig-arg">homologs</span>=<span class="sig-default">True</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ifr.interfaces.gather-pysrc.html#gather.query">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Processes the query for the given list of genes. The result will be to
  cache the results in self.data. Use other methods of this class to query 
  the cached data, for example, to get the data for annotation GO:0009615, 
  which is 'response to virus', use getDataForAnnotation()</p>
  <dl class="fields">
    <dt>Parameters:</dt>
    <dd><ul class="nomargin-top">
        <li><strong class="pname"><code>genelist</code></strong> - A list of gene strings like ['RSAD2', 'SERPING1',....]</li>
        <li><strong class="pname"><code>homologs</code></strong> - If True, then the query includes annotations for homologs to the 
          input genelist, else it does not.</li>
    </ul></dd>
    <dt>Returns:</dt>
        <dd>A return code indicating status of request. The data will be 
          stored in self.data[] as a list of lines.</dd>
  </dl>
</td></tr></table>
</div>
<a name="getDataForAnnotation"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">getDataForAnnotation</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">key</span>=<span class="sig-default"><code class="variable-quote">'</code><code class="variable-string">GO:0009615</code><code class="variable-quote">'</code></span>,
        <span class="sig-arg">return_mult</span>=<span class="sig-default">True</span>,
        <span class="sig-arg">bf_thresh</span>=<span class="sig-default">0.0</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ifr.interfaces.gather-pysrc.html#gather.getDataForAnnotation">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Returns the parsed data for all rows which contain a given key, which 
  can be any string that one expects to find in the response data from a 
  gather query.</p>
  <dl class="fields">
    <dt>Parameters:</dt>
    <dd><ul class="nomargin-top">
        <li><strong class="pname"><code>key</code></strong> - A string, which could be a gene ontology id, a word (such as 
          'virus'), or any other text that can be found in the result set.</li>
        <li><strong class="pname"><code>return_mult</code></strong> - Set true to return multiple rows which meet the search key. False
          finds only the first row that has the search key.</li>
        <li><strong class="pname"><code>bf_thresh</code></strong> - The bayes factor threshold so that only annotations meeting or 
          exceeding this threshold will be returned.</li>
    </ul></dd>
    <dt>Returns:</dt>
        <dd>A tuple (fields, dat), where fields provides the header data 
          defining the fields of the returned data, and dat is a list of 
          lists. Each top-level list is from a row of the response that 
          contains the key string. The second level is the list of field 
          values for each row, where the field values correspond to the 
          'fields' header information.</dd>
  </dl>
<div class="fields">      <p><strong>Note:</strong>
        You must have called the query() method to load the gather data for
        the query genes prior to using this function.
      </p>
</div></td></tr></table>
</div>
<a name="getBayesFactor"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">getBayesFactor</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">row_id</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ifr.interfaces.gather-pysrc.html#gather.getBayesFactor">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Given the gather row identifier (not the index of self.data, but 
  rather the first field of a response row in the data), returns the bayes 
  factor and p-value.</p>
  <dl class="fields">
    <dt>Parameters:</dt>
    <dd><ul class="nomargin-top">
        <li><strong class="pname"><code>row_id</code></strong> - The row identifier, which should be an integer.</li>
    </ul></dd>
    <dt>Returns:</dt>
        <dd>(label, bayes factor, p-value). See the Gather documentation for 
          an explanation.</dd>
  </dl>
<div class="fields">      <p><strong>Note:</strong>
        self.data must be populated by a call to query() before this method
        will work.
      </p>
</div></td></tr></table>
</div>
<a name="parseAnnotation"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">parseAnnotation</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">annotationList</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ifr.interfaces.gather-pysrc.html#gather.parseAnnotation">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Parses an annotation to generate a dictionary with key-value 
  bindings.</p>
  <dl class="fields">
    <dt>Returns:</dt>
        <dd>a dictionary with key/value bindings describing the fields of a 
          given annotation.</dd>
  </dl>
</td></tr></table>
</div>
<br />
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="ifr-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="module-tree.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            >Iterative Feature Removal</th>
          </tr></table></th>
  </tr>
</table>
<table border="0" cellpadding="0" cellspacing="0" width="100%%">
  <tr>
    <td align="left" class="footer">
    Generated by Epydoc 3.0.1 on Tue Sep 17 11:02:40 2013
    </td>
    <td align="right" class="footer">
      <a target="mainFrame" href="http://epydoc.sourceforge.net"
        >http://epydoc.sourceforge.net</a>
    </td>
  </tr>
</table>

<script type="text/javascript">
  <!--
  // Private objects are initially displayed (because if
  // javascript is turned off then we want them to be
  // visible); but by default, we want to hide them.  So hide
  // them unless we have a cookie that says to show them.
  checkCookie();
  // -->
</script>
</body>
</html>
